packages/notifications/push/amplify_push_notifications/ios/Classes/AmplifyPushNotificationsPlugin+AppDelegate.swift

```diff
@@ -0,0 +1,120 @@
+                launchNotification = remoteNotification
+
+                // NOTE: the notification payload will also be passed into didReceiveRemoteNotification below after
+                // this delegate method, didFinishLaunchingWithOptions completes.
+                // As this notification is recorded as the luanch notification, it should not be sent via
+                // sharedEventsStreamHandlers.notificationOpened, this check is handled in didReceiveRemoteNotification.
+            }
+
+            // Otherwise the host App is launched in the background, this notification will be sent to Flutter
+            // via flutterApi.onNotificationReceivedInBackground in didReceiveRemoteNotification below.
+
+            // After the host App launched in the background, didFinishLaunchingWithOptions will no longer
+            // be fired when an end user taps a notification.
+
+            // After the host App launched in the background, it runs developers' Flutter code as well including
+            // Amplify.addPlugin and Amplify.confogure, and the listeners of notification events should be
+            // attached. When an end user taps a notification from this point, the tapped notification will be
+            // sent via sharedEventsStreamHandlers.notificationOpened in didReceiveRemoteNotification below.
+        }
+
+        return true
+    }
+
+    // Method is being called differently at different states of App life cycle
+    // https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1623013-application#discussion
+    // Specifically:
+    // 1. .background case triggers when:
+    //   a. remote notifications background mode is enabled on the host App AND -
+    //   b. remote notifications has `content-avilable: 1` AND -
+    //   c. the host App is in the background
+    // 2. .inactive case triggers when:
+    //   a. end user taps on a notification to open the host App regardless of whether the App was terminated
+    //      or was in the background
+    // 3. .active case triggers when:
+    //   a. a remote notification arrives when the App is in the foreground
+    public func application(
+        _ application: UIApplication,
+        didReceiveRemoteNotification userInfo: [AnyHashable : Any],
+        fetchCompletionHandler completionHandler: @escaping (UIBackgroundFetchResult) -> Void
+    ) -> Bool {
+        switch UIApplication.shared.applicationState {
+        case .background:
+            flutterApi.onNotificationReceivedInBackground(withPayload: userInfo) { _ in
+                // Wait for Flutter side to complete calling registered callbacks
+                // then notifiy the system that the background task is completed
+                completionHandler(.noData)
+            }
+        case .inactive:
+
+            if let launchNotification = launchNotification {
+                if !NSDictionary(dictionary: launchNotification).isEqual(to: userInfo) {
```

Comment:
I think this just compares references, though, right? Does this work?

Reply:
`isEqual` does a deep comparison

>Two dictionaries have equal contents if they each hold the same number of entries and, for a given key, the corresponding value objects in each dictionary satisfy the [isEqual(_:)](https://developer.apple.com/documentation/objectivec/nsobjectprotocol/1418795-isequal) test.